aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,158 @@
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/UnitTests/HomeControllerIndexPost.cs
+  :language: c#
+  :emphasize-lines: 14-15,20,27-28,31,38
+
+The first test confirms that, when the ``ModelState`` is not valid, the method returns the same ``ViewResult`` as a GET request would. Note that the test doesn't attempt to pass in an invalid model. That wouldn't work anyway since model binding isn't running - we're just calling the method directly. However, we're not trying to test model binding - we're only testing what our code in the action method does. So the simplest thing is to just explicitly add an error to ``ModelState``.
+
+The second test verifies that when ``ModelState`` is valid, a new ``BrainStormSession`` is added (via the repository), and the method returns a ``RedirectToActionResult`` with the expected properties. Mocked calls that aren't called are normally ignored, but calling ``Verifiable`` at the end of the setup call allows it to be verified in the test. This is done with the call to ``mockRepo.Verify``.
+
+Another controller in the app displays information related to a particular brainstorming session. It includes some logic to deal with invalid id values:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Controllers/SessionController.cs
+  :language: c#
+  :emphasize-lines: 16,20,25,33
+
+Looking at this controller action, it should be clear that there are at least three cases to test, one for each ``return`` statement. The unit tests are shown below:
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/UnitTests/SessionControllerIndex.cs
+  :language: c#
+  :emphasize-lines: 16,26,39
+
+Finally, the application exposes some functionality as a web API, including a list of ideas associated with a brainstorming session and a method for adding new ideas to a session. The controller is shown below:
+
+.. _ideas-controller:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Api/IdeasController.cs
+  :language: c#
+  :emphasize-lines: 20-22,27,29-35,49-51,55,60,70
+
+The ``ForSession`` method returns a list of ``dynamic`` types, with property names camel cased to match JavaScript conventions. Avoid returning your business domain entities directly via API calls, since frequently they include more data than the API client requires, and they unnecessarily couple your app's internal domain model with the API you expose externally. You can define strongly typed data-transfer objects (DTOs), or just return dynamic as this method does. Mapping between domain entities and the types you will return over the wire can be done manually (using a LINQ ``Select`` as shown here) or using a library like `AutoMapper <https://github.com/AutoMapper/AutoMapper>`_
```

Comment:
These are *not* (!!!!!!) dynamic types! These are *anonymous* types. There's a huge difference between these and they actually have nothing in common (in part, because there's no such thing as a "dynamic type" :smile:).